File: client/native/usb.js
// native/usb.js — WebUSB (request) glue for the Tier-3 seam.
//
// `navigator.usb.requestDevice({filters})` opens the OS chooser and returns a
// `USBDevice`. The device is not JSON-able, so we hold it in a module-level
// registry keyed by an incrementing id and hand the id plus identifying fields
// back across the wire.

import { CapabilityError } from "./index.js";

/**
 * Live USB devices, keyed by the id handed back to Python.
 * @type {Map<number, any>}
 */
const _devices = new Map();

/** Monotonic id counter — stable in tests (never Math.random/Date). */
let _counter = 0;

/**
 * Whether the WebUSB API is available.
 * @param {Object} _args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{supported:boolean}>}
 */
export async function usbIsSupported(_args, deps) {
  const nav = deps.navigator || /** @type {any} */ (globalThis).navigator;
  return { supported: !!(nav && nav.usb) };
}

/**
 * Request a USB device and register it.
 * @param {{filters?:Array<Object>}} args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{id:number, vendor_id:number, product_id:number, product_name:string}>}
 * @throws {CapabilityError} unavailable / cancelled / failed.
 */
export async function usbRequest(args, deps) {
  const nav = deps.navigator || /** @type {any} */ (globalThis).navigator;
  if (!nav || !nav.usb || typeof nav.usb.requestDevice !== "function") {
    throw new CapabilityError("unavailable", "the WebUSB API is not available");
  }
  let device;
  try {
    device = await nav.usb.requestDevice({ filters: args.filters || [] });
  } catch (err) {
    if (err && err.name === "AbortError") {
      throw new CapabilityError("cancelled", err.message);
    }
    throw new CapabilityError("failed", err && err.message);
  }
  _counter += 1;
  const id = _counter;
  _devices.set(id, device);
  return {
    id,
    vendor_id: device.vendorId,
    product_id: device.productId,
    product_name: device.productName || "",
  };
}

File: tempestweb/native/usb.py
"""Native USB capability over the WebUSB API.

:func:`request` sends a ``usb.request`` ``native_call`` and gets back a device
carrying an opaque id plus identifying fields; ``client/native/usb.js`` calls
``navigator.usb.requestDevice`` with the supplied filters and stores the live
``USBDevice`` in a registry keyed by that id. The Python side never touches the
device — it only shuttles the id and metadata.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from tempestweb.native.dispatch import send_native_call

__all__ = ["UsbDevice", "is_supported", "request"]


@dataclass(frozen=True)
class UsbDevice:
    """A USB device granted through the WebUSB API.

    Attributes:
        id: The opaque device id; the client holds the live ``USBDevice`` in a
            registry keyed by this id.
        vendor_id: The USB vendor id.
        product_id: The USB product id.
        product_name: The device's product name, or ``""`` when unknown.
    """

    id: str
    vendor_id: int
    product_id: int
    product_name: str


async def is_supported() -> bool:
    """Report whether the WebUSB API is available.

    Returns:
        ``True`` if the browser exposes ``navigator.usb``.

    Raises:
        BrowserUnavailableError: If called with no native bridge installed.
    """
    value = await send_native_call("usb.is_supported", {})
    return bool(value.get("supported", False))


async def request(filters: list[dict[str, Any]] | None = None) -> UsbDevice:
    """Request a USB device through the browser chooser.

    Args:
        filters: The ``USBDeviceFilter`` dicts constraining the chooser; an empty
            list offers all devices.

    Returns:
        The granted :class:`UsbDevice`.

    Raises:
        NativeError: If the API is unavailable (``unavailable``) or the chooser is
            dismissed/refused (``permission_denied``).
        BrowserUnavailableError: If called with no native bridge installed.
    """
    filters = filters or []
    value = await send_native_call("usb.request", {"filters": filters})
    return UsbDevice(
        id=str(value.get("id", "")),
        vendor_id=int(value.get("vendor_id", 0)),
        product_id=int(value.get("product_id", 0)),
        product_name=str(value.get("product_name", "")),
    )
